MSVC: Windows compatibility

jhunkeler opened the pull request:
Resolves [SCSB-201](https://jira.stsci.edu/browse/SCSB-201)

This PR makes it possible to compile `stcal` using MSVC/MSBuild on Windows.


- MSVC does not support Variable Length Arrays ([ref](https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-170#c-standards-support-1))
- POSIX resource management functions do not exist on Windows (`sys/resource.h`, etc.).
  - If need be, we can implement the same functionality using the [Windows API](https://learn.microsoft.com/en-us/windows/win32/psapi/collecting-memory-usage-information-for-a-process)

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

jhunkeler committed:
Replaces variable length array in is_pix_in_list()

jhunkeler committed:
Prevent compiler error if NPY_NTYPES* symbols are undefined

jhunkeler committed:
Adjust preprocessor directives for Windows compat.

jhunkeler committed:
MSVC wants macros wrapped

jhunkeler committed:
Disable memory usage output

jhunkeler committed:
MSVC: ERROR is reserved by the compiler

jhunkeler committed:
Replace variable length arrays

jhunkeler committed:
Use GetProcessMemory on Windows

jhunkeler committed:
Link to PSAPI library on Windows

jhunkeler committed:
Fix broken stage

jhunkeler committed:
Fix broken stage

pllim commented:
Should add a Windows build here to prove this PR works. 😉 

https://github.com/spacetelescope/stcal/blob/583930c60aaa44fa9b72ac739752a0a36235ea89/.github/workflows/ci.yml#L31

mcara reviewed (approved):
Amazing!

jhunkeler committed:
Define PATH_MAX only when necessary

jhunkeler committed:
GHA: Add Windows targets

jhunkeler commented:
Of course the moment I decide to add Windows into the mix:
https://github.com/python/cpython/issues/135151

jhunkeler committed:
GHA: Skip over 3.13.4 due to bad library linkage

jhunkeler committed:
Fix target name

pllim commented:
You finally get to the fun stuff:

`assert str(arr._filename).split("/")[-1] in os.listdir(tempdir)`

and

```
>       return io.open(self, mode, buffering, encoding, errors, newline)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       PermissionError: [Errno 13] Permission denied: 'C:\\...\\test_disk_appendable_array_bad0\\tmptest'
```

jhunkeler committed:
Align directory open exeception with Linux/Mac:

* On Windows opening a directory as a file results in a "PermissionError" exception
* Here we trap the exception, and if we're on Windows, check whether the file name is a directory. If so raise an "IsADirectoryError" exception

jhunkeler committed:
Replace complex call

* "fname" gives us the absolute path to the file, so there's no need to
  list the destination directory to see if the file is in there.
  fname.exists will suffice.

jhunkeler committed:
Fix bad method calls

jhunkeler commented:
Tada.
https://github.com/spacetelescope/stcal/actions/runs/15644177990/job/44078328011?pr=351

pllim commented on tests/outlier_detection/test_median.py:
```diff
@@ -24,7 +24,7 @@ def test_disk_appendable_array(tmp_path):
 
     # check temporary file setup
-    assert str(arr._filename).split("/")[-1] in os.listdir(tempdir)  # noqa: SLF001
+    assert fname.exists() # noqa: SLF001
```
```suggestion
    assert fname.exists()
```

pllim commented:
Maybe this warrants a change log so Windows user can read it and be happy.

jhunkeler committed:
Update tests/outlier_detection/test_median.py

Removes noqa comment

Co-authored-by: P. L. Lim <2090236+pllim@users.noreply.github.com>

pllim commented:
FWIW green job is good enough for merge. We can always follow-up later if there is any Windows intricacies to fix later (I think that is going to get smoke out if downstream also start doing regression tests in Windows 👹). But since I am not maintainers, I'll leave approval and merge to someone else. Thanks!
